File: tree.go
package main

// 二叉树
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// createBinaryTreeFromSlice creates a binary tree from a slice of integers
// Assuming the slice is completely filled with integers, starting with the root at index 0
// func createBinaryTreeFromSlice(slice []interface{}) *TreeNode {
// 	if len(slice) == 0 {
// 		return nil
// 	}
// 	if slice[0] == nil {
// 		return nil
// 	}
// 	root := &TreeNode{Val: slice[0].(int)}

//		// If there's only one element, return the root
//		if len(slice) == 1 {
//			return root
//		}
//		// Recursively create the left and right subtrees
//		root.Left = createBinaryTreeFromSlice(slice[1 : (len(slice)+1)/2])
//		root.Right = createBinaryTreeFromSlice(slice[(len(slice)+1)/2:])
//		return root
//	}

// 需要适配LeetCode的二叉树数组表示形式！
// @author Bingo
func sliceToBinaryTree(nums []interface{}) *TreeNode {
	if len(nums) == 0 || nums[0] == nil {
		return nil
	}

	root := &TreeNode{Val: nums[0].(int)}
	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(nums) {
		node := queue[0]
		queue = queue[1:]

		if nums[i] != nil {
			node.Left = &TreeNode{Val: nums[i].(int)}
			queue = append(queue, node.Left)
		}
		i++

		if i < len(nums) && nums[i] != nil {
			node.Right = &TreeNode{Val: nums[i].(int)}
			queue = append(queue, node.Right)
		}
		i++
	}

	return root
}

// @author 文心一言
// traverseBinaryTree traverses the binary tree and returns a slice of int or nil values
func traverseBinaryTree(root *TreeNode) []interface{} {
	if root == nil {
		return nil
	}

	// Create a slice to store the values and nils
	result := make([]interface{}, 0)

	// Preorder traversal: root -> left -> right
	result = append(result, root.Val)
	result = append(result, traverseBinaryTree(root.Left)...)
	result = append(result, traverseBinaryTree(root.Right)...)

	// Check if the left child is nil and the right child is not nil
	// If so, insert a nil value to represent the null left child
	if root.Left == nil && root.Right != nil {
		result = append(result, nil)
	}

	return result
}

// @author Copilot
func createTree(arr []interface{}) *Node {
	if len(arr) == 0 || arr[0] == nil {
		return nil
	}

	root := &Node{Val: arr[0].(int)}
	queue := []*Node{root}
	i := 2

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		for i < len(arr) && arr[i] != nil {
			child := &Node{Val: arr[i].(int)}
			node.Children = append(node.Children, child)
			queue = append(queue, child)
			i++
		}

		i++
	}

	return root
}

// ========================
// N叉树
type Node struct {
	Val      int
	Children []*Node
}

// @author Bingo
func sliceToNTree(nums []interface{}) *Node {
	if len(nums) == 0 || nums[0] == nil {
		return nil
	}

	root := &Node{Val: nums[0].(int)}
	queue := []*Node{root}
	i := 1

	for len(queue) > 0 && i < len(nums) {
		node := queue[0]
		queue = queue[1:]

		for i < len(nums) && nums[i] != nil {
			child := &Node{Val: nums[i].(int)}
			node.Children = append(node.Children, child)
			queue = append(queue, child)
			i++
		}
		if i < len(nums) && nums[i] == nil {
			i++ // Skip the nil marking the end of a line
		}
	}

	return root
}

// @author 文心一言
// convertSliceToNaryTree converts a slice of integers or nils to an N-ary tree
func convertSliceToNaryTree(slice []interface{}) *Node {
	if len(slice) == 0 {
		return nil
	}

	// Create the root node
	root := &Node{Val: slice[0].(int)}

	// Iterate over the rest of the slice to build the tree
	children := make([]*Node, 0)
	for i := 1; i < len(slice); i++ {
		// Check if the current element is nil, indicating a new level
		if slice[i] == nil {
			// If children are present, add them to the previous node
			if len(children) > 0 {
				prevNode := children[len(children)-1]
				prevNode.Children = children
				children = make([]*Node, 0) // Reset children slice for the next level
			}
		} else {
			// Create a new child node
			newNode := &Node{Val: slice[i].(int)}
			children = append(children, newNode)
		}
	}

	// If there are remaining children, set them as the last node's children
	if len(children) > 0 {
		lastNode := children[len(children)-1]
		lastNode.Children = children
	}

	return root
}

File: trees/tree.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# @author 文心一言
def slice_to_binary_tree(nums):
    if not nums or nums[0] is None:
        return None

    root = TreeNode(nums[0])
    queue = [root]
    i = 1

    while queue and i < len(nums):
        node = queue.pop(0)

        if nums[i] is not None:
            node.left = TreeNode(nums[i])
            queue.append(node.left)
        i += 1

        if i < len(nums) and nums[i] is not None:
            node.right = TreeNode(nums[i])
            queue.append(node.right)
        i += 1

    return root

# @author 文心一言
def is_same_tree(p: TreeNode, q: TreeNode) -> bool:
    # 如果两个节点都为空，则它们是相等的
    if p is None and q is None:
        return True
    # 如果一个节点为空，而另一个不为空，则它们不相等
    if p is None or q is None:
        return False
    # 如果节点的值不相等，则它们不相等
    if p.val != q.val:
        return False
    # 递归地检查左子树和右子树是否相等
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

# @author 文心一言
def buildTree(preorder, postorder):
    if not preorder or not postorder:
        return None

    # 后序遍历的最后一个节点是根节点
    root_val = postorder[-1]
    root = TreeNode(root_val)

    # 在前序遍历中找到根节点的索引
    root_index = preorder.index(root_val)

    # 前序遍历中根节点左边的部分对应后序遍历中的左子树
    # 前序遍历中根节点右边的部分对应后序遍历中的右子树
    left_preorder = preorder[1:root_index + 1]
    right_preorder = preorder[root_index + 1:]
    left_postorder = postorder[:root_index]
    right_postorder = postorder[root_index:-1]

    # 递归构造左子树和右子树
    root.left = buildTree(left_preorder, left_postorder)
    root.right = buildTree(right_preorder, right_postorder)

    return root

# 如果左子树为空则用None占位，但是如果右子树为空则不需要None
# @author Bingo
def serializeBinaryTree(root):
    def dfs(node):
        if node is None:
            return []
        left = dfs(node.left)
        right = dfs(node.right)
        return [node.val] + ([None] if (not left and right) else left) + right
    return dfs(root)

# 还需要再调整一下，只要遇到空节点就用None占位，末尾的所有None可以省略
# @author Bingo
def serializeBinaryTree_2(root):
    def dfs(node):
        if node is None:
            return [None]
        left = dfs(node.left)
        right = dfs(node.right)
        return [node.val] + left + right
    result = dfs(root)
    while result and result[-1] is None:
        result.pop()
    return result

# 再调整一下，从前序遍历改为改为层序遍历，遇到空则放None
# @author Bingo
def level_order_traversal(root):
    if root is None:
        return []
    queue = deque([root])
    result = []
    while queue:
        node = queue.popleft()
        if node is None:
            result.append(None)
            continue
        result.append(node.val)
        # if node.left is not None or node.right is not None:
        queue.append(node.left)
        queue.append(node.right)
    while result and result[-1] is None:
        result.pop()
    return result
